test/dlc_tests/conftest.py

```diff
@@ -291,15 +305,211 @@ def ec2_instance_role_name(request):
+        if region == "us-east-1"
+        else UBUNTU_18_BASE_DLAMI_US_WEST_2
+    )
 
 
 @pytest.fixture(scope="function")
 def ei_accelerator_type(request):
     return request.param if hasattr(request, "param") else None
 
 
+@pytest.mark.timeout(300)
+@pytest.fixture(scope="function")
+def efa_ec2_instances(
+    request,
+    ec2_client,
+    ec2_resource,
+    ec2_instance_type,
+    ec2_instance_role_name,
+    ec2_key_name,
+    ec2_instance_ami,
+    region,
+    availability_zone_options,
+):
+    ec2_key_name = f"{ec2_key_name}-{str(uuid.uuid4())}"
+    print(f"Creating instance: CI-CD {ec2_key_name}")
+    key_filename = test_utils.generate_ssh_keypair(ec2_client, ec2_key_name)
+
+    def delete_ssh_keypair():
+        if test_utils.is_pr_context():
+            test_utils.destroy_ssh_keypair(ec2_client, key_filename)
+        else:
+            with open(KEYS_TO_DESTROY_FILE, "a") as destroy_keys:
+                destroy_keys.write(f"{key_filename}\n")
+
+    request.addfinalizer(delete_ssh_keypair)
+
+    response = None
+    instance_name_prefix = f"CI-CD {ec2_key_name}"
+    for availability_zone in availability_zone_options:
+        try:
+            response = ec2_client.run_instances(
+                BlockDeviceMappings=[
+                    {
+                        "DeviceName": "/dev/sda1",
+                        "Ebs": {
+                            "DeleteOnTermination": True,
+                            "VolumeSize": 150,
+                            "VolumeType": "gp3",
+                            "Iops": 3000,
+                            "Throughput": 125,
+                        },
+                    },
+                ],
+                ImageId=ec2_instance_ami,
+                InstanceType=ec2_instance_type,
+                IamInstanceProfile={"Name": ec2_instance_role_name},
+                KeyName=ec2_key_name,
+                MaxCount=2,
+                MinCount=2,
+                NetworkInterfaces=ec2_utils.generate_network_interfaces(
+                    ec2_client, ec2_instance_type, availability_zone
+                ),
+                Placement={"AvailabilityZone": availability_zone},
+                TagSpecifications=[
+                    {
+                        "ResourceType": "instance",
+                        "Tags": [{"Key": "Name", "Value": instance_name_prefix}],
+                    }
+                ],
+            )
+            if response and response["Instances"]:
+                break
+        except ClientError as e:
+            LOGGER.warning(
+                f"Failed to launch in {availability_zone} due to {e}\n"
+                "Retrying in the next availability zone."
+            )
+            continue
+    if not (response and response["Instances"]):
+        raise RuntimeError(f"Unable to launch {ec2_instance_type} instances in {region}")
+
+    instances = response["Instances"]
+
+    def terminate_efa_instances():
```

Comment:
do we delete key pairs as well

Reply:
Yes, refer to line 355 in this file, where we add a finalizer to delete the keypair.